### PR TITLE
App parts

### LINKED ISSUE #14509: Create 3.0 version of App parts
*  Look at sample code. Create new samples with 3.0 templates using same namespace.
*  Copy over supporting code from previous version and test
* Create whole-topic versions - that is

```
::: moniker range=">= aspnetcore-3.0"
Copy paste entire topic here
::: moniker-end
::: moniker range="< aspnetcore-3.0"
Current topic here
::: moniker-end
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2784a724-e940-4c14-3b50-b063f78c2ee5
* Version Independent ID: 41175c3a-8cb5-856b-a7e0-931edb6394db
* Content: [Application Parts in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-2.2&branch=pr-en-us-14507)
* Content Source: [aspnetcore/mvc/advanced/app-parts.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/advanced/app-parts.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### LINKED ISSUE #14593: RazorViewEngineOptions for FileProviders in ASP.NET Core 3.0
Moved from https://github.com/aspnet/AspNetCore.Docs/issues/14586 (EDIT by @guardrex to place code block into remark).

@zhou881117 remarks ...

> ```csharp
> services.Configure(options =>
> {
>     options.FileProviders.Add(
>         new EmbeddedFileProvider(typeof(MySharedController).GetTypeInfo().Assembly));
> });
> ```
> .net core 3.0 RazorViewEngineOptions changed. without FileProviders to locate view
> 
> @Rick-Anderson
> tks!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8d8787c1-1a02-ae68-aa59-7c0e2e823dde
* Version Independent ID: 3e799e5d-a89c-cd03-999f-92ae431f4657
* Content: [File Providers in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/file-providers?view=aspnetcore-3.0)
* Content Source: [aspnetcore/fundamentals/file-providers.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/fundamentals/file-providers.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #15078: [MVC] Include a section about discovery in the application part docs
In this doc
https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0

We get a lot of issues regarding 404s when invoking controllers or pages with all of them traced back to the user missing some key piece on the way application parts get discovered and work.

We want to add an additional section on the document above to explain the application part discovery process and mention the requirements for MVC applications to work outside usage patterns.

Specifically:
* The `applicationName` setting needs to be set to the root assembly (normally the entry point assembly) used for discovery.
* The root assembly needs to have a reference to the parts used for discovery (either direct or transitive).
* The root assembly needs to reference the Web SDK.
  * We have custom build logic that stamps attributes into the root assembly that are used for discovery.

@pranavkm Anything missing here?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2784a724-e940-4c14-3b50-b063f78c2ee5
* Version Independent ID: 41175c3a-8cb5-856b-a7e0-931edb6394db
* Content: [Application Parts in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-2.2&branch=pr-en-us-14507)
* Content Source: [aspnetcore/mvc/advanced/app-parts.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/advanced/app-parts.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #14593
Fixes #15078
Fixes #14509
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0&branch=pr-en-us-15593)

Upgraded to 3.0 and tested. Only these things to review:

* [feature providers](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0&branch=pr-en-us-15593#feature-providers)
* [Discovery in application parts](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0&branch=pr-en-us-15593#discovery-in-application-parts)
* **Include views**  To include views in the assembly: Add the `EmbeddedFileProvider ` to the RazorViewEngine: -- Startup & Startup2 don't have the `EmbeddedFileProvider `  and they display the view. Did that change in 2.x?

### CODE REVIEW

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/StartupViews.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public IConfiguration Configuration { get; }
+
+        #region snippet
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<MvcRazorRuntimeCompilationOptions>(options =>
+            {
+                options.FileProviders.Add(   // <-
```
Comment (Rick-Anderson) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/StartupViews.cs]: @serpent5  here's the problem.

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/MySharedApp/MySharedApp.csproj [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+    </Content>
+  </ItemGroup>
+
+  <ItemGroup>
+    <EmbeddedResource Include="Views\**\*.cshtml" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.App" />
```
Comment (serpent5) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/MySharedApp/MySharedApp.csproj]:
```suggestion
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
```

This is what's causing the warning in the Visual Studio solution explorer.

**Comment by guardrex** [comment]
I broke this :boom: with that patch on the versioning, so I resolved the conflict to set it back into a good state.

**Comment by Rick-Anderson** [comment]
@pranavkm 
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0&branch=pr-en-us-15593)

Upgraded to 3.0 and tested. Only these things to review:

* [feature providers](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0&branch=pr-en-us-15593#feature-providers)
* [Discovery in application parts](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/advanced/app-parts?view=aspnetcore-3.0&branch=pr-en-us-15593#discovery-in-application-parts)
* **Include views**  To include views in the assembly: Add the `EmbeddedFileProvider ` to the RazorViewEngine: -- Startup & Startup2 don't have the `EmbeddedFileProvider `  and they display the view. Did that change in 2.x? Can I remove Add the `EmbeddedFileProvider `?

**Comment by Rick-Anderson** [comment]
@pranavkm can you review or suggest a reviewer?

**Comment by Rick-Anderson** [comment]
@pranavkm can you review or suggest a reviewer - or give me an ETA?

File: aspnetcore/mvc/advanced/app-parts.md [hunk 3]
```diff
@@ -126,6 +134,15 @@ View Components:
...
 ```
 
+## Discovery in application parts
+
+HTTP 404 errors are not uncommon when developing with application parts. These errors are typically caused by missing an essential requirement for how applications parts are discovered. If your app returns an HTTP 404 error, verify the following requirements have been met:
+
+* The `applicationName` setting needs to be set to the root assembly used for discovery. The root assembly used for discovery is normally the entry point assembly.
+* The root assembly needs to have a reference to the parts used for discovery. The reference can be direct or transitive.
+* The root assembly needs to reference the Web SDK.
+  * The ASP.NET Core framework has custom build logic that stamps attributes into the root assembly that are used for discovery.
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts.md]: Formatting

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/MySharedApp/MySharedApp.csproj [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Content Include="Views\MyShared\Index.cshtml">
+      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+    </Content>
+  </ItemGroup>
+
+  <ItemGroup>
+    <EmbeddedResource Include="Views\**\*.cshtml" />
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/MySharedApp/MySharedApp.csproj]: We do not want to encourage users to use embedded views. Can this be changed to a regular RCL?

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Program.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+            CreateHostBuilder(args).Build().Run();
+        }
+
+        public static IHostBuilder CreateHostBuilder(string[] args) =>
+            Host.CreateDefaultBuilder(args)
+                .ConfigureWebHostDefaults(webBuilder =>
+                {
+                    webBuilder
+                     .UseStartup<Startup>();
+                    // .UseStartup<Startup2>();
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Program.cs]: ?

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            Configuration = configuration;
+        }
+
+        public IConfiguration Configuration { get; }
+
+        #region snippet
+        // Requires using System.Reflection;
+        public void ConfigureServices(IServiceCollection services)
+        {
+            var assembly = typeof(MySharedController).GetTypeInfo().Assembly;
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup.cs]:
```suggestion
            var assembly = typeof(MySharedController).Assembly;
```

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup2.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using MySharedApp.Controllers;
+using System.Reflection;
+
+
+namespace WebAppParts
+{
+    public class Startup2
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup2.cs]: How is different from `Startup`?

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/StartupRm.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            services.AddControllersWithViews()
+                    .ConfigureApplicationPartManager(apm =>
+                    {
+                        var dependentLibrary = apm.ApplicationParts
+                            .FirstOrDefault(part => part.Name == "MySharedApp");
+
+                        if (dependentLibrary != null)
+                        {
+                            apm.ApplicationParts.Remove(dependentLibrary);
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/StartupRm.cs]: Can we remove this sample? I'm not sure what value there is in showing users ways to remove things from application discovery.

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/StartupViews.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        #region snippet
+        // requires 
+        // using Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation;
+        // using System.Reflection;
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<MvcRazorRuntimeCompilationOptions>(options =>
+            {
+                options.FileProviders.Add(
+                 new EmbeddedFileProvider(typeof(MySharedController).GetTypeInfo().Assembly));
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/StartupViews.cs]: Let's not document this. RCLs should be a sufficient replacement for this (which the first sample covers).

File: aspnetcore/mvc/advanced/app-parts/3.0sample2/AppPartsSample/GenericControllerFeatureProvider.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    // An IApplicationFeatureProvider<ControllerFeature> is used at startup time to
+    // discover types that should be treated as controllers. ASP.NET Core does not regster
+    // open generic type controllers.
+    //
+    // This component creates a new type (e.g. GenericController<Widget>) for each entity 
+    // type that does not have a non-generic controller defined. This is determined based 
+    // on the type name. Because SprocketController is defined, GenericController<Sprocket> 
+    // is not added as it would create a conflict.
+    #region snippet
+    public class GenericControllerFeatureProvider :
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample2/AppPartsSample/GenericControllerFeatureProvider.cs]: Can we remove this? We wouldn't recommend doing this to our users.

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Program.cs (hunk 5 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Program.cs]: I have two different startups for showing different features. They each need to be a snippet. This is a typical pattern we follow in the docs.

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup2.cs (hunk 7 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup2.cs]: Startup.cs
```
public void ConfigureServices(IServiceCollection services)
{
    var assembly = typeof(MySharedController).GetTypeInfo().Assembly;
    services.AddControllersWithViews()
        .AddApplicationPart(assembly);
}
```

Startup2.cs
```
// Requires using System.Reflection;
// Requires using Microsoft.AspNetCore.Mvc.ApplicationParts;
public void ConfigureServices(IServiceCollection services)
{
    var assembly = typeof(MySharedController).GetTypeInfo().Assembly;
    var part = new AssemblyPart(assembly);
    services.AddControllersWithViews()
        .ConfigureApplicationPartManager(apm => apm.ApplicationParts.Add(part));
}
```

**Comment by Rick-Anderson** [comment]
> Can you give me a gist of what these samples are meant to demonstrate? Some of these such as the ones that remove a part or implement a custom feature provider for generic types are patterns we wouldn't recommend to users.

I have no idea. I'll delete your suggested delete and change to RCL.

File: aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup2.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+        public IConfiguration Configuration { get; }
+
+        #region snippet
+        // Requires using System.Reflection;
+        // Requires using Microsoft.AspNetCore.Mvc.ApplicationParts;
+        public void ConfigureServices(IServiceCollection services)
+        {
+            var assembly = typeof(MySharedController).GetTypeInfo().Assembly;
+            var part = new AssemblyPart(assembly);
```
Comment (pranavkm) [inline aspnetcore/mvc/advanced/app-parts/3.0sample1/WebAppParts/Startup2.cs]:
```suggestion
            // This creates an AssemblyPart, but does not create any related parts for items such as views.
            var part = new AssemblyPart(assembly);
```

**Comment by luminalpark** [comment]
Hi,

testing this example I get an error 500 when calling /MyShared/IndexView

The view 'Index' was not found. The following locations were searched: /Views/MyShared/Index.cshtml /Views/Shared/Index.cshtml /Pages/Shared/Index.cshtml

Tried adding

`services.AddRazorPages().AddRazorRuntimeCompilation();
            services.Configure<MvcRazorRuntimeCompilationOptions>(options =>
            {
                options.FileProviders.Add(
                    new EmbeddedFileProvider(assembly));
            });`

Issue is still there.
Can you have a look?

(Using .Net Core 3.0)

**Comment by gsitv** [comment]
Hi, I tried your sample with core 3.0. same like @luminalpark  I get an error. Seems like embeded razor views are not found.

Can you resolve this please?

**Comment by jlKampos** [comment]
Same issue reported by @gsitv  and @luminalpark, since there is no solution a work around is to move all the views to the web app project, have not tested, but it brakes the principle so ...
Any ideais  ?

**Comment by pranavkm** [comment]
Hi folks, please don't respond to a closed PR since it's very easy to miss seeing these notifications. If you think there's a bug in the framework, please file an issue in https://github.com/dotnet/aspnetcore with a minimal app that reproduces the issue. Thanks!

**Pull request merged**